plugin/agent/workloadattestor-k8s/k8s.go

```diff
@@ -0,0 +1,188 @@
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/hcl"
+	"github.com/spiffe/spire/proto/agent/workloadattestor"
+	"github.com/spiffe/spire/proto/common"
+	spi "github.com/spiffe/spire/proto/common/plugin"
+)
+
+type k8sPlugin struct {
+	kubeletReadOnlyPort int
+	httpClient          httpClient
+	fs                  fileSystem
+	mtx                 *sync.RWMutex
+}
+
+type k8sPluginConfig struct {
+	KubeletReadOnlyPort int `hcl:"kubelet_read_only_port"`
+}
+
+type pod struct {
+	// We only care about namespace, serviceAccountName and containerID
+	Metadata struct {
+	} `json:"metadata"`
+	Items []struct {
+		Metadata struct {
+			Namespace string `json:"namespace"`
+		} `json:"metadata"`
+		Spec struct {
+			ServiceAccountName string `json:"serviceAccountName"`
+		} `json:"spec"`
+		Status struct {
+			ContainerStatuses []struct {
+				ContainerID string `json:"containerID"`
+			} `json:"containerStatuses"`
+		} `json:"status"`
+	} `json:"items"`
+}
+
+const (
+	selectorType string = "k8s"
+)
+
+func (p *k8sPlugin) Attest(req *workloadattestor.AttestRequest) (*workloadattestor.AttestResponse, error) {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	log.Printf("Attesting PID: %v", req.Pid)
+	resp := workloadattestor.AttestResponse{}
+
+	httpResp, err := p.httpClient.Get(fmt.Sprintf("http://localhost:%v/pods", p.kubeletReadOnlyPort))
+	if err != nil {
+		return &resp, err
+	}
+	defer httpResp.Body.Close()
+	respBytes, err := ioutil.ReadAll(httpResp.Body)
+
+	var podInfo *pod
+	err = json.Unmarshal(respBytes, &podInfo)
+	if err != nil {
+		return &resp, err
+	}
+
+	cgroups, err := getCgroups(fmt.Sprintf("/proc/%v/cgroup", req.Pid), p.fs)
+	if err != nil {
+		return &resp, err
+	}
+
+	var containerID string
+	for _, cgroup := range cgroups {
+		// We are only interested in kube pods entries
+		if len(cgroup[2]) < 9 {
+			continue
+		}
+
+		substring := cgroup[2][:9]
+		if substring == "/kubepods" {
+			parts := strings.Split(cgroup[2], "/")
+
+			if len(parts) < 5 {
+				log.Printf("Kube pod entry found, but without container id: %v", substring)
+				continue
+			}
+			containerID = parts[4]
+			break
+		}
+	}
+
+	for _, item := range podInfo.Items {
```

Comment:
if the pid wasn't in a pod, or we couldn't find it's container id, we could skip this part and bail out early

Reply:
The parsing of cgroups information will be done first now, so this will not be reached if the PID was not in a pod. And if the response from the API does not include pod items, it will just skip the loop.